### ISSUE TITLE
Allow asynchronous check before ApplePay sheet is shown

### ISSUE THREAD

**Issue by phuhl** [Adyen/adyen-web#1801]
**Is your feature request related to a problem? Please describe.**
When any kind of async check needs to be performed (e.g. we need to do an API call to check for conflicts) when the user clicks on the Apple Pay button, with the current implementation the Apple Pay sheet won't open due the "Must create a new ApplePaySession from a user gesture handler" limitation.

I thought for a moment, that I could use the "onClick" method, but it is awaited **before** the ApplePaySession is created. So this is not possible for this use case.

**Describe the solution you'd like**
An optional async callback between `new ApplePayService` and `session.begin();` in `src/components/ApplePay/ApplePay.tsx` would resolve my issue. E.g. a function `confirmSession` that can be passed in the `create` method:

```diff
    private startSession(onPaymentAuthorized) {
        const { version, onValidateMerchant, onCancel, onPaymentMethodSelected, onShippingMethodSelected, onShippingContactSelected } = this.props;

        return new Promise((resolve, reject) => this.props.onClick(resolve, reject)).then(() => {
            const paymentRequest = preparePaymentRequest({
                companyName: this.props.configuration.merchantName,
                ...this.props
            });

            const session = new ApplePayService(paymentRequest, { /* ... */ });
+           this.props.confirmSession && await this.props.confirmSession();
            session.begin();
        });
    }
```

**Describe alternatives you've considered**
I think, due to the limitations of the Apple Pay API, my use case can only be done this way.

**Additional context**
The way I propose the feature, no existing APIs are broken and I know from experience with another PSP that the approach I suggest is viable (Apple Pay does allow an await before the `session.begin` call). That also matches Apples documentation, which describes the limitation only for the creation of the ApplePaySession object (https://developer.apple.com/documentation/apple_pay_on_the_web/apple_pay_js_api/creating_an_apple_pay_session) but not for the begin method (https://developer.apple.com/documentation/apple_pay_on_the_web/applepaysession/1778001-begin).

![image](https://user-images.githubusercontent.com/16835755/196741085-d8781b8a-bc4e-4691-8cde-946369d01755.png)

**Comment by phuhl** [comment]
A possible alternative would be to change how onClick behaves:

```ts
    private startSession(onPaymentAuthorized) {
        const { version, onValidateMerchant, onCancel, onPaymentMethodSelected, onShippingMethodSelected, onShippingContactSelected } = this.props;

        const paymentRequest = preparePaymentRequest({
            companyName: this.props.configuration.merchantName,
            ...this.props
        });

        const session = new ApplePayService(paymentRequest, { /* ... */ });

        return new Promise((resolve, reject) => this.props.onClick(resolve, reject)).then(() => {
            session.begin();
        });
    }
```

This would be API wise the nicest option, but has the downside of changing current behavior.

**Comment by ribeiroguilherme** [comment]
Hey @phuhl. Thanks for reaching out, and sorry for getting back to you just now.

I agree with you that second approach would be the best one, but indeed that would lead to a breaking change. 
It seems like the first way you proposed is the one that would not affect the current implementation. We could provide a callback like `onBeforeStartSession` (or something similar).

We created a ticket in our backlog to investigate this further and work on a fix. Out of curiosity, did you find a workaround for now?

For internal reference: FOC-67196

**Comment by ribeiroguilherme** [comment]
@phuhl we released the fix for this on version 5.39.0 . You can use the `onClick`callback for performing the async check. Cheers
